File: 2017/day22/day22.py
#!/usr/bin/env python3
# [Day 22: Sporifica Virus](https://adventofcode.com/2017/day/22)

import sys
from collections import defaultdict
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
self_tests = "-T" in sys.argv
if self_tests:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()

nx = len(lines[0])
ny = len(lines)


MOVES = ((0, -1), (1, 0), (0, 1), (-1, 0))  # up/right/down/left


# part 1

infected = set()
for y, line in enumerate(lines):
    for x, c in enumerate(line):
        if c == "#":
            infected.add((x, y))

x, y = nx // 2, ny // 2  # middle of the map
d = 0  # facing up

infections1 = 0
for _ in range(10000):
    if (x, y) in infected:
        d = (d + 1) % 4  # turn right
        infected.discard((x, y))  # infect the node
    else:
        d = (d + 3) % 4  # turn left
        infected.add((x, y))  # infect the node
        infections1 += 1

    dx, dy = MOVES[d]  # move
    x, y = x + dx, y + dy

print(infections1)


# part 2

CLEAN = 0
WEAKENED = 1
INFECTED = 2
FLAGGED = 3

nodes = defaultdict(lambda: CLEAN)
for y, line in enumerate(lines):
    for x, c in enumerate(line):
        if c == "#":
            nodes[x, y] = INFECTED

x, y = nx // 2, ny // 2  # middle of the map
d = 0  # facing up

infections2 = 0
for _ in range(10_000_000):
    node = nodes[x, y]

    if node == CLEAN:
        d = (d + 3) % 4  # turn left
        node = WEAKENED

    elif node == WEAKENED:
        node = INFECTED
        # do not turn
        infections2 += 1

    elif node == INFECTED:
        d = (d + 1) % 4  # turn right
        node = FLAGGED

    elif node == FLAGGED:
        node = CLEAN
        d = (d + 2) % 4  # go back

    else:
        assert False

    nodes[x, y] = node

    dx, dy = MOVES[d]  # move
    x, y = x + dx, y + dy

print(infections2)


# test
if filename == "test":
    assert infections1 == 5587
    assert infections2 == 2511944

File: 2017/day22/day22.rs
//! [Day 22: Sporifica Virus](https://adventofcode.com/2017/day/22)

use rustc_hash::{FxHashMap, FxHashSet};

const MOVES: &[(i32, i32)] = &[
    (0, -1), // up
    (1, 0),  // right
    (0, 1),  // down
    (-1, 0), // left
];

/// Node states for part 2
#[derive(Clone, Copy)]
enum State {
    Clean,
    Weakened,
    Infected,
    Flagged,
}

struct Puzzle {
    infected: FxHashSet<(i32, i32)>,
    nx: i32, // size of the map
    ny: i32,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        let mut infected = FxHashSet::default();
        let mut nx = 0;
        let mut ny = 0;

        for (y, line) in (0..).zip(data.lines()) {
            for (x, c) in (0..).zip(line.chars()) {
                if c == '#' {
                    infected.insert((x, y));
                }
                nx = nx.max(x + 1);
            }
            ny = y + 1;
        }

        Self { infected, nx, ny }
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        let mut infected = self.infected.clone();

        let mut x = self.nx / 2; // middle of the map
        let mut y = self.ny / 2;
        let mut dir: usize = 0; // start facing up

        let mut infections = 0;

        for _ in 0..10_000 {
            if infected.remove(&(x, y)) {
                dir = (dir + 1) % 4; // turn right and clean the node
            } else {
                dir = (dir + 3) % 4; // turn left
                infected.insert((x, y)); // infect the node
                infections += 1;
            }

            // move
            let (dx, dy) = MOVES[dir];
            x += dx;
            y += dy;
        }

        infections
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        let mut nodes = self
            .infected
            .iter()
            .copied()
            .map(|(x, y)| ((x, y), State::Infected))
            .collect::<FxHashMap<(i32, i32), State>>();

        let mut x = self.nx / 2; // middle of the map
        let mut y = self.ny / 2;
        let mut dir: usize = 0; // start facing up

        let mut infections = 0;

        for _ in 0..10_000_000 {
            let node = nodes.entry((x, y)).or_insert(State::Clean);

            (dir, *node) = match node {
                State::Clean => ((dir + 3) % 4, State::Weakened), // turn left
                State::Weakened => {
                    infections += 1;
                    (dir, State::Infected) // do not turn
                }
                State::Infected => ((dir + 1) % 4, State::Flagged), // turn right
                State::Flagged => ((dir + 2) % 4, State::Clean),    // go back
            };

            // move
            let (dx, dy) = MOVES[dir];
            x += dx;
            y += dy;
        }

        infections
    }
}

fn main() {
    let args = aoc::parse_args();
    let puzzle = Puzzle::new(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part1() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.part1(), 5587);
    }

    #[test]
    fn part2() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.part2(), 2511944);
    }
}
